src/libraries/System.Net.Security/src/System/Net/Security/NetSecurityTelemetry.cs

```diff
@@ -0,0 +1,307 @@
+using System.Threading;
+using Microsoft.Extensions.Internal;
+
+namespace System.Net.Security
+{
+    [EventSource(Name = "System.Net.Security")]
+    internal sealed class NetSecurityTelemetry : EventSource
+    {
+        public static readonly NetSecurityTelemetry Log = new NetSecurityTelemetry();
+
+        private IncrementingPollingCounter? _tlsHandshakeRateCounter;
+        private PollingCounter? _totalTlsHandshakesCounter;
+        private PollingCounter? _currentTlsHandshakesCounter;
+        private PollingCounter? _failedTlsHandshakesCounter;
+        private PollingCounter? _sessionsOpenCounter;
+        private PollingCounter? _sessionsOpenTls10Counter;
+        private PollingCounter? _sessionsOpenTls11Counter;
+        private PollingCounter? _sessionsOpenTls12Counter;
+        private PollingCounter? _sessionsOpenTls13Counter;
+        private EventCounter? _handshakeDurationCounter;
+        private EventCounter? _handshakeDurationTls10Counter;
+        private EventCounter? _handshakeDurationTls11Counter;
+        private EventCounter? _handshakeDurationTls12Counter;
+        private EventCounter? _handshakeDurationTls13Counter;
+
+        private long _finishedTlsHandshakes; // Successfully and failed
+        private long _startedTlsHandshakes;
+        private long _failedTlsHandshakes;
+        private long _sessionsOpen;
+        private long _sessionsOpenTls10;
+        private long _sessionsOpenTls11;
+        private long _sessionsOpenTls12;
+        private long _sessionsOpenTls13;
+
+        protected override void OnEventCommand(EventCommandEventArgs command)
+        {
+            if (command.Command == EventCommand.Enable)
+            {
+                _tlsHandshakeRateCounter ??= new IncrementingPollingCounter("tls-handshake-rate", this, () => Interlocked.Read(ref _finishedTlsHandshakes))
+                {
+                    DisplayName = "TLS handshakes completed",
+                    DisplayRateTimeScale = TimeSpan.FromSeconds(1)
+                };
+
+                _totalTlsHandshakesCounter ??= new PollingCounter("total-tls-handshakes", this, () => Interlocked.Read(ref _finishedTlsHandshakes))
+                {
+                    DisplayName = "Total TLS handshakes completed"
+                };
+
+                _currentTlsHandshakesCounter ??= new PollingCounter("current-tls-handshakes", this, () => -Interlocked.Read(ref _finishedTlsHandshakes) + Interlocked.Read(ref _startedTlsHandshakes))
+                {
+                    DisplayName = "Current TLS handshakes"
+                };
+
+                _failedTlsHandshakesCounter ??= new PollingCounter("failed-tls-handshakes", this, () => Interlocked.Read(ref _failedTlsHandshakes))
+                {
+                    DisplayName = "Total TLS handshakes failed"
+                };
+
+                _sessionsOpenCounter ??= new PollingCounter("all-sessions-open", this, () => Interlocked.Read(ref _sessionsOpen))
```

Comment:
```suggestion
                _sessionsOpenCounter ??= new PollingCounter("tls-sessions-open", this, () => Interlocked.Read(ref _sessionsOpen))
```
'all-sessions' seems ambiguous - I assume there are many other kinds of authenticated communication sessions even if .NET doesn't implement them (Kerberos, SSL, SSH?). Everywhere else we used 'tls' so can we use it here too?

Reply:
Here it includes tls and ssl2, ssl3. (same for comments below).

Using tls in that case would be less misleading than all. I'm not aware of a different term that would be more appropriate here.

Reply:
> Using tls in that case would be less misleading than al

In context I am guessing you meant **more** misleading? : ) What about "TLS/SSL Sessions Active"?

Reply:
> What about "TLS/SSL Sessions Active"?

As the description, yes. I suppose that's enough clarification even if we use the `tls-*` event name.

Reply:
tls-ssl-sessions-open?

Reply:
That does drag along all "generic" counters:
`tls-handshakes-per-second`
`total-tls-handshakes`
`current-tls-handshakes`
`failed-tls-handshakes`

to
`tls-ssl-handshakes-per-second`
`total-tls-ssl-handshakes`
`current-tls-ssl-handshakes`
`failed-tls-ssl-handshakes`

@wfurt Do you have a preference here regarding "tls-" vs "tls-ssl-" vs "all-" naming?

Reply:
I would use "tls-". We use either one but I have not seem them together like this. Possibly something like "all-tls-sessions-open".
To @noahfalk 's point. We may add some telemetry to NegotiatStream/Kerberos.